flink-core/src/main/java/org/apache/flink/core/fs/FileSystem.java

```diff
@@ -671,16 +672,111 @@ public boolean exists(final Path f) throws IOException {
+        }
+
+        final FileStatus[] containingFiles = listStatus(path);
+        if (containingFiles == null) {
+            LOG.warn(
+                    "No files could be retrieved even though the path was marked as existing. This is an indication for a bug in the underlying FileSystem implementation and should be reported. It won't affect the processing of this run, though.");
+            return;
+        }
+
+        IOException exception = null;
+        for (FileStatus fileStatus : containingFiles) {
+            final Path childPath = fileStatus.getPath();
+
+            if (childPath.equals(path)) {
+                // we need to make sure that we delete the path itself after all its content is
+                // deleted
+                continue;
+            }
+
+            try {
+                deleteRecursively(childPath);
+            } catch (IOException e) {
+                exception = ExceptionUtils.firstOrSuppressed(e, exception);
+            }
+        }
+
+        if (exception == null) {
+            deleteFile(path);
+        } else {
+            throw exception;
+        }
+    }
+
+    /**
+     * Deletes the object referenced by the passed {@code path}.
+     *
+     * @param path The path referring to the object that shall be deleted.
+     * @throws java.nio.file.DirectoryNotEmptyException if the {@code path} refers to a non-empty
+     *     directory.
+     * @throws IOException if an error occurred while deleting the file other than the {@code path}
+     *     referring to a non-empty directory.
+     */
+    private void deleteFile(Path path) throws IOException {
+        if (isDirectoryWithContent(path)) {
+            throw new DirectoryNotEmptyException(path.getPath());
+        }
+
+        try {
+            tryToDelete(path);
+        } catch (IOException e) {
+            if (exists(path)) {
+                throw e;
+            }
+            return;
+        }
+
+        if (exists(path)) {
+            throw new IOException(path.getPath() + " could not be deleted for unknown reasons.");
+        }
+    }
+
+    /**
+     * Checks whether a given {@code path} refers to a non-empty directory.
+     *
+     * @param path The path referring to the directory in question.
+     * @return {@code true}, if the {@code path} refers to a non-empty directory; otherwise {@code
+     *     false}.
+     * @throws IOException If an error occurred while accessing the underlying file system.
+     */
+    protected abstract boolean isDirectoryWithContent(Path path) throws IOException;
+
+    /**
+     * Deletes the file or empty directory referenced by the given {@code path}. Shouldn't do
+     * anything if the object doesn't exist.
+     *
+     * @param path The path to the object that is subject for deletion.
+     * @throws IOException If an error occurs while accessing the underlying file system.
      */
-    public abstract boolean delete(Path f, boolean recursive) throws IOException;
```

Comment:
I do not think you should delete the old method. The class is annotated `@Public` and can only change with a new Flink major version. It would also be great to see a FLIP for this change because changes to the public API require generally a FLIP.

Reply:
Fair point. I didn't delete it, though. The interface still matches. Just the implementation changed. That would work, wouldn't it? 🤔

Reply:
...the JavaDoc was made more specific, though, which could be considered a change in contract.

Reply:
Ah missed that it is not deleted but I guess it is not intended to be overwritten anymore and users are instead urged to overwrite `tryToDelete`. One idea to keep the backwards compatibility would be to implement an empty `tryDelete` and keep other FileSystems overwrite `delete`.

Reply:
Summary of an offline discussion: We might want to move this change further down the class hierarchy to not affect the public interface `FileSystem` that much.